compiler/rustc_target/src/target_features.rs

```diff
@@ -15,45 +17,105 @@ pub const RUSTC_SPECIFIC_FEATURES: &[&str] = &["crt-static"];
+    /// nightly and using it in `#[target_feature]` requires enabling the given nightly feature.
+    Unstable {
+        /// This must be a *language* feature, or else rustc will ICE when reporting a missing
+        /// feature gate!
+        nightly_feature: Symbol,
+        /// See `Stable::allow_toggle` comment above.
+        allow_toggle: AllowToggle,
+    },
+    /// This feature can not be set via `-Ctarget-feature` or `#[target_feature]`, it can only be
+    /// set in the basic target definition. It is never set in `cfg(target_feature)`. Used in
+    /// particular for features that change the floating-point ABI.
     Forbidden { reason: &'static str },
 }
-use Stability::*;
 
-impl<CTX> HashStable<CTX> for Stability {
+/// `Stability` where `allow_toggle` has not been computed yet.
+/// Returns `Ok` if the toggle is allowed, `Err` with an explanation of not.
+pub type StabilityUncomputed = Stability<fn(&Target) -> Result<(), &'static str>>;
+/// `Stability` where `allow_toggle` has already been computed.
+pub type StabilityComputed = Stability<Result<(), &'static str>>;
+
+impl<CTX, AllowToggle: HashStable<CTX>> HashStable<CTX> for Stability<AllowToggle> {
     #[inline]
     fn hash_stable(&self, hcx: &mut CTX, hasher: &mut StableHasher) {
         std::mem::discriminant(self).hash_stable(hcx, hasher);
         match self {
-            Stable => {}
-            Unstable(sym) => {
-                sym.hash_stable(hcx, hasher);
+            Stability::Stable { allow_toggle } => {
+                allow_toggle.hash_stable(hcx, hasher);
+            }
+            Stability::Unstable { nightly_feature, allow_toggle } => {
+                nightly_feature.hash_stable(hcx, hasher);
+                allow_toggle.hash_stable(hcx, hasher);
             }
-            Forbidden { .. } => {}
+            Stability::Forbidden { reason } => {
+                reason.hash_stable(hcx, hasher);
+            }
+        }
+    }
+}
+
+impl<AllowToggle> Stability<AllowToggle> {
+    /// Returns whether the feature can be queried in `cfg` ever.
+    /// (It might still be nightly-only even if this returns `true`).
+    pub fn in_cfg(self) -> bool {
+        !matches!(self, Stability::Forbidden { .. })
+    }
+
+    /// Returns the nightly feature that is required to toggle or query this target feature. Ensure
+    /// to also check `allow_toggle()` before allowing to toggle!
+    pub fn requires_nightly(self) -> Option<Symbol> {
+        match self {
+            Stability::Unstable { nightly_feature, .. } => Some(nightly_feature),
+            Stability::Stable { .. } => None,
+            Stability::Forbidden { .. } => panic!("forbidden features should not reach this far"),
         }
     }
 }
 
-impl Stability {
-    pub fn is_stable(self) -> bool {
-        matches!(self, Stable)
+impl StabilityUncomputed {
+    pub fn compute(self, target: &Target) -> StabilityComputed {
```

Comment:
can this be, like, `compute_toggleability`?

Reply:
:shrug: sure. Seems like the type aliases should also have a different name then, but `StabilityWithComputedToggleability` is a bit too much IMO...

Reply:
Yeah I think we can leave the nouns as-is, I just want this particular verb to be a bit clearer.

Reply:
It's renamed in the latest diff.